src/main/java/io/netty/incubator/codec/http3/Http3FrameCodec.java

```diff
@@ -721,6 +723,69 @@ private void unsetState(int toUnset) {
         }
     }
 
+    private static final class WriteResumptionListener
+            implements GenericFutureListener<Future<? super QuicStreamChannel>> {
+        private static final Object FLUSH = new Object();
+        private final Deque<Object> buffer;
+        private final ChannelHandlerContext ctx;
+        private final Http3FrameCodec codec;
+
+        WriteResumptionListener(ChannelHandlerContext ctx, Http3FrameCodec codec) {
+            this.ctx = ctx;
+            this.codec = codec;
+            buffer = new ArrayDeque<>(4); // assuming we will buffer header, data, trailer and a flush
+            assert codec.qpackAttributes != null;
+            codec.qpackAttributes.whenEncoderStreamAvailable(this);
+        }
+
+        @Override
+        public void operationComplete(Future<? super QuicStreamChannel> future) {
+            drain();
+        }
+
+        void enqueue(Object msg, ChannelPromise promise) {
```

Comment:
what happens if this method would be called after `operationComplete(...)` was called ? Should we keep track and call `drain()` ?

Reply:
The logic is a bit spread out but it is not possible for `enqueue` to be called after `operationComplete()` returns as we set `codec.writeResumptionListener` to `null` after we are done draining. Additionally, since we drain the queue in a loop, re-entrant writes also are drained.